### DISCUSSION TITLE
Explanation about decltype used in typechecker_spec.lua

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#88] [Q&A] [answered]
In the following lines https://github.com/edubart/nelua-lang/blob/e6727dd704438dea658a049bb1315627286fe459/spec/typechecker_spec.lua#L1814-L1816 we see 

```lua
local x: auto = 1 
```

and then `x` is used inside `decltype`:

```lua
local y: decltype(x) = x
```

Does this mean `y`'s `decltype` behavior is like C++14's `decltype(auto)` in this case?

**Comment by edubart** [comment, marked as answer]
The `decltype` in Nelua is a generic that returns the type of an identifier or expression.

> Does this mean y's decltype behavior is like C++14's decltype(auto) in this case?

Not exactly, Nelua's `decltype` is similar to C++'s `decltype` because you can use both to get a type from an identifier or expression, but `decltype(auto)` in C++ is used only to automatic detect the type for function returns while in Nelua the same is done just with `auto` or leaving the function returns without type notations.

Here is a small example of decltype usage:

```lua
local function bounded_increment(x: auto): auto
  local y: decltype(x) = x + 1
  if y <= x then -- value wrapped around due to integer overflow
    y = x
  end
  return y
end

print(bounded_increment(0.0)) -- outputs: 1.0
print(bounded_increment(4294967295_u32)) -- outputs: 4294967295
```

So `decltype` can be used enforce a type early, but usually this is not needed as the compiler can infer types by itself, so it's very little used, it's rare to need it.

**Reply by stefanos82** [reply]
Thank you for the explanation @edubart.

Coming from C and C++, using `__typeof__`, `auto`, and `decltype(auto)` can make you see such behavior in Nelua's `auto` and `decltype`, thus the mild confusion.
